apsuite/optics_analysis/orbit_correction.py

```diff
@@ -0,0 +1,178 @@
+        if corridx is None:
+            corridx = self.corr_idx
+        if delta_kicks is None:
+            raise Exception('Missing Delta Kicks values')
+        for idx_mag, mag in enumerate(corridx):
+            delta = delta_kicks[idx_mag]/len(mag)
+            for _, seg in enumerate(mag):
+                if idx_mag < self._nch:
+                    kick = model[seg].hkick_polynom + delta
+                    model[seg].hkick_polynom = kick
+                else:
+                    kick = model[seg].vkick_polynom + delta
+                    model[seg].vkick_polynom = kick
+
+    def _check_kicks(self, model=None, chidx=None, cvidx=None):
+        if model is None:
+            model = self.model
+        if chidx is None:
+            chidx = self.ch_idx
+        if cvidx is None:
+            cvidx = self.cv_idx
+
+        kicks = self.get_kicks(model, chidx, cvidx)
+        kicksx, kicksy = kicks[:self._nch], kicks[self._nch:]
+        kicksx_above = kicksx > OrbitCorr.MAX_HKICK
+        kicksy_above = kicksy > OrbitCorr.MAX_VKICK
+
+        if sum(kicksx_above) or sum(kicksy_above):
+            str_above = '\n'
+            str_above += 'HKick > MaxHKick at CHs: {0:s} \n'
+            str_above += 'VKick > MaxVKick at CVs: {1:s}'
+            xlist = str(np.argwhere(kicksx_above).flatten())
+            ylist = str(np.argwhere(kicksy_above).flatten())
+            raise Exception(str_above.format(xlist, ylist))
+
+    @staticmethod
+    def get_figm(res):
+        """Calculate figure of merit from residue vector."""
+        return np.sqrt(np.sum(res*res)/res.size)
+
+    def orbit_corr(self,
+                   model,
+                   jacobian_matrix=None,
+                   goal_orbit=None,
+                   nsv=None, nr_max=10, tol=1e-6):
+        """Orbit correction.
+
+        Calculates the pseudo-inverse of orbit correction matrix via SVD
+        and minimizes the residue vector [CODx@BPM, CODy@BPM].
+        """
+        if goal_orbit is None:
+            goal_orbit = np.zeros(2 * self._nbpm)
+
+        if jacobian_matrix is None:
+            jmat = self.calc_jacobian_matrix(model)
+        else:
+            jmat = _dcopy(jacobian_matrix)
+
+        if jmat.shape[1] > self._ncorr:
+            jmat = jmat[:, :-1]
+
+        umat, smat, vmat = np.linalg.svd(jmat, full_matrices=False)
+        ismat = 1/smat
+        ismat[np.isnan(ismat)] = 0
+        ismat[np.isinf(ismat)] = 0
+        if nsv is not None:
+            ismat[nsv:] = 0
+        ismat = np.diag(ismat)
+        ismat = -1 * np.dot(np.dot(vmat.T, ismat), umat.T)
+        orb = self._get_orbit_residue(model)
+        dorb = orb - goal_orbit
+        bestfigm = OrbitCorr.get_figm(dorb)
+        if bestfigm < tol:
+            return OrbitCorr.CORR_STATUS.Sucess
+
+        for _ in range(nr_max):
+            dkicks = np.dot(ismat, dorb)
+            self._set_delta_kicks(model=model, delta_kicks=dkicks)
+            self._check_kicks(model=model)
```

Comment:
you are applying the kicks and then checking them against the maximum values, but you don't remove the delta applied before raising the error. Isn't it better to remove or don't even apply the exceeding kicks?

Reply:
when I was implementing I thought about this question and here is what I decided:

I had the "let the user handle it" policy in my mind. If the kicks are greater than the maximum the user must restart the correction with a different singular values or corrector selection, beginning with the initial kicks again. 

The chance that before the kicks surpasses the max value they are already close to the max is high, then keeping the almost spoiled model may not be very useful. 

In the case that the variations are big and suddenly the kicks overcomes the max, it is an indicative that the singular values should be reduced and the correction must restart anyway.

what do you think?

Reply:
Applying the kicks and then checking them makes easier (at least to me) to get the list of corrector whose kicks are greater than max.

Reply:
ok! lets do this way then!